File: src/main/java/org/example/Greedy.java
package org.example;

import java.util.ArrayList;
import java.util.List;

public class Greedy {

    public static List<Integer> greedyVC(int[][] inputGraph) {
        List<Integer> cover = new ArrayList<>();
        boolean valid;
        int[] numEdge;

        do {
            numEdge = countNumEdge(inputGraph);
            int m = getMaxNumEdgeIndex(numEdge);
            cover.add(m);
            valid = isValidCover(inputGraph, cover);
            System.out.println("s");
        } while (!valid);

        System.out.println(cover);
        var c = cover;
        return cover;
    }

    private static int[] countNumEdge(int[][] graph) {
        int n = graph.length;
        int[] numEdge = new int[n];

        for (int i = 0; i < n; i++) {
            for (int j = i; j < n; j++) {
                if (graph[i][j] == 1) {
                    numEdge[i]++;
                    numEdge[j]++;
                }
            }
        }

        return numEdge;
    }

    private static int getMaxNumEdgeIndex(int[] numEdge) {
        int maxIndex = 0;
        int maxNumEdge = 0;

        for (int i = 0; i < numEdge.length; i++) {
            if (numEdge[i] > maxNumEdge) {
                maxNumEdge = numEdge[i];
                maxIndex = i;
            }
        }

        return maxIndex;
    }

    private static boolean isValidCover(int[][] graph, List<Integer> cover) {
        boolean valid = true;
        int n = graph.length;
        int[] numEdge = new int[n];

        for (int i = 0; i < n; i++) {
            for (int j = i; j < n; j++) {
                if (graph[i][j] == 1) {
                    if (!cover.contains(i) && !cover.contains(j)) {
                        valid = false;
                        numEdge[i]++;
                        numEdge[j]++;
                    }
                }
            }
        }

        return valid;
    }

/*
    //возвращает индекс вершины с наибольшей степенью
    private static int getMaxDeg(int[] vertex) {
        int I = 0;
        int max = 0;

        for (int i = 0; i < vertex.length; i++) {
            if (vertex[i] > max) {
                max = vertex[i];
                I = i;
            }
        }

        return I;
    }

    //меняет значения в массиве степеней у вершин
    private static int[] countVertexDegs(int[][] graph) {
        int n = graph.length;
        int[] vertex_deg = new int[n];

        for(int i = 0; i < n; i++) {
            vertex_deg[i] = 0;

            for(int j = 0; j < n; j++) {
                if(graph[i][j] == 1) {
                    vertex_deg[i]++;
                    if(i == j) vertex_deg[i]++;
                }
            }
        }

        return vertex_deg;
    }

    //в графе все нули или нет?
    public static boolean isEmpty(int[] vertex) {
        for(int i = 0; i < vertex.length; i++) {
            if (vertex[i] != 0) {
                return false;
            }
        }

        return true;
    }

    //решение
    public static List<Integer> solve(int[][] igraph) {
        var graph = (int[][]) igraph.clone();
        List<Integer> result = new ArrayList<>();

        int n = graph.length;

        int[] vertex_deg = countVertexDegs(graph);

        int vertex;

        for(; result.size() < n && !isEmpty(vertex_deg);) {
            //жадно берем вершину с наибольшей степенью
            vertex = getMaxDeg(vertex_deg);
            result.add(vertex);

            //удаляем эту вершину из графа и из массива степеней
            for(int i = 0; i < n; i++) {
                graph[vertex][i] = graph[i][vertex] = 0;
            }

            vertex_deg = countVertexDegs(graph);
        }

        return result;
    }*/
}

File: algorythm/greedy.py
import random


def get_max_deg(vertex):
    I = 0
    max_deg = 0
    for i in range(len(vertex)):
        if vertex[i] > max_deg:
            max_deg = vertex[i]
            I = i
    return I


def count_vertex_degs(graph):
    n = len(graph)
    vertex_deg = [0] * n

    for i in range(n):
        for j in range(n):
            if graph[i][j] == 1:
                vertex_deg[i] += 1
                if i == j:
                    vertex_deg[i] += 1
    return vertex_deg


def is_empty(vertex):
    return all(v == 0 for v in vertex)


def solve(graph):
    result = []

    n = len(graph)

    vertex_deg = count_vertex_degs(graph)

    while len(result) < n and not is_empty(vertex_deg):
        vertex = get_max_deg(vertex_deg)
        result.append(vertex)

        for i in range(n):
            graph[vertex][i] = graph[i][vertex] = 0

        vertex_deg = count_vertex_degs(graph)

    return result


def generate_random_adjacency_matrix(n):
    adjacency_matrix = [[random.randint(0, 1)
                         for _ in range(n)] for _ in range(n)]

    for i in range(n):
        if adjacency_matrix[i][i] == 1:
            adjacency_matrix[i][i] = 0

    for i in range(n):
        for j in range(i + 1, n):
            adjacency_matrix[j][i] = adjacency_matrix[i][j]

    return adjacency_matrix
